File: tests/st/ops/dynamic_shape/grad/test_squeeze.py
# Copyright 2022 Huawei Technologies Co., Ltd
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ============================================================================

import numpy as np
import pytest
from mindspore import nn
from mindspore.ops import operations as P
from mindspore import context
from mindspore import Tensor
from .test_grad_of_dynamic import TestDynamicGrad


class NetSqueeze(nn.Cell):
    def __init__(self):
        super(NetSqueeze, self).__init__()
        self.squeeze = P.Squeeze()

    def construct(self, x):
        return self.squeeze(x)


@pytest.mark.level1
@pytest.mark.platform_x86_gpu_training
@pytest.mark.platform_x86_cpu
@pytest.mark.env_onecard
def test_dynamic_squeeze_shape():
    """
    Feature: Squeeze Grad DynamicShape.
    Description: Test case of dynamic shape for Squeeze grad operator on CPU, GPU and Ascend.
    Expectation: success.
    """
    context.set_context(mode=context.PYNATIVE_MODE)
    test_dynamic = TestDynamicGrad(NetSqueeze())
    x = Tensor(np.array([[-0.1, 0.3, 3.6], [0.4, 0.5, -3.2]]).astype(np.float16))
    test_dynamic.test_dynamic_grad_net(x)


@pytest.mark.level1
@pytest.mark.platform_x86_gpu_training
@pytest.mark.platform_x86_cpu
@pytest.mark.env_onecard
def test_dynamic_squeeze_rank():
    """
    Feature: Squeeze Grad DynamicRank.
    Description: Test case of dynamic rank for Squeeze grad operator on CPU, GPU and Ascend.
    Expectation: success.
    """
    context.set_context(mode=context.PYNATIVE_MODE)
    test_dynamic = TestDynamicGrad(NetSqueeze())
    x = Tensor(np.array([[-0.1, 0.3, 3.6], [0.4, 0.5, -3.2]]).astype(np.float16))
    test_dynamic.test_dynamic_grad_net(x, True)

File: tests/ut/cpp/ops/view/test_squeeze.cc
/**
 * Copyright 2023 Huawei Technologies Co., Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
#include "test_view.h"
#include "mindspore/core/ops/view/squeeze_strides_calc.h"

namespace mindspore {
namespace ops {
class TestViewSqueeze : public TestView {
 public:
  TestViewSqueeze() {}
};

/// Feature: squeeze strides calculator
/// Description: Test view squeeze strides calculator is right
/// Expectation: success
TEST_F(TestViewSqueeze, View) {
  auto prim = std::make_shared<Primitive>("Squeeze");
  std::vector<int64_t> tensor_data = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
  auto input_tensor = std::make_shared<tensor::Tensor>(tensor_data, kInt64);
  input_tensor->set_shape({2, 1, 1, 5});

  std::vector<int64_t> axis_data = {2};
  auto input_axis = MakeValue(axis_data);

  // test size
  std::vector<ValuePtr> inputs_squeeze_size;
  inputs_squeeze_size.push_back(input_tensor);
  inputs_squeeze_size.push_back(input_axis);

  ASSERT_TRUE(SqueezeCalc(prim, inputs_squeeze_size).empty());

  // test nullptr
  std::vector<ValuePtr> inputs_squeeze_null;
  auto nullinput_tensor = nullptr;
  inputs_squeeze_null.push_back(nullinput_tensor);
  ASSERT_TRUE(SqueezeCalc(prim, inputs_squeeze_null).empty());

  // test type
  std::vector<ValuePtr> inputs_squeeze_type;
  inputs_squeeze_type.push_back(input_axis);
  ASSERT_TRUE(SqueezeCalc(prim, inputs_squeeze_type).empty());

  std::vector<ValuePtr> inputs_squeeze;
  inputs_squeeze.push_back(input_tensor);
  prim->AddAttr(kAxis, input_axis);
  auto storage_info = SqueezeCalc(prim, inputs_squeeze);
  std::vector<int64_t> expect_shape({2, 1, 5});

  ASSERT_FALSE(storage_info.empty());
  ASSERT_TRUE(storage_info[0]->is_contiguous);
  ASSERT_TRUE(storage_info[0]->shape == expect_shape);

  axis_data = {};
  input_axis = MakeValue(axis_data);
  prim->AddAttr(kAxis, input_axis);
  storage_info = SqueezeCalc(prim, inputs_squeeze);
  std::vector<int64_t> expect_shape_2({2, 5});
  ASSERT_TRUE(storage_info[0]->shape == expect_shape_2);
}

}  // namespace ops
}  // namespace mindspore
